### DISCUSSION TITLE
Feature Request: Configure DuckDB through environment variables

### DISCUSSION THREAD

**Discussion by janjagusch** [duckdb/duckdb#9845] [Ideas]
I want to be able to configure DuckDB through environment variables. For example, instead of running:

```
SET custom_extension_repository='asdf.example';
```

At the beginning of every DuckDB session, I want to specify this through an environment variable,:

```
export DUCKDB_CUSTOM_EXTENSION_REPOSITORY='asf.examlpe'
```

Generally, I suggest that all options should be configurable through `DUCKDB_`-prefixed environment variables. This is already possible for some options, but not all, for example:
https://github.com/duckdb/duckdb/blob/3f96a90635d351089a0214b8f1d11b7c7244e3bd/extension/httpfs/include/s3fs.hpp#L26-L27

This is especially relevant when you work with other tooling in the DuckDB ecosystem, like [Rill](https://github.com/rilldata/rill), where you have no direct access to the CLI (https://github.com/rilldata/rill/issues/3576).

**Comment by Mytherin** [comment]
You can do this manually in the CLI using the `getenv` function since https://github.com/duckdb/duckdb/pull/9651. We don't support it in other clients for security reasons - although we could perhaps ship the `getenv` function in an extension.

**Reply by janjagusch** [reply]
Thanks for your reply @Mytherin. I think there has been a misunderstanding. I don't want to source environment variables in the CLI. I want to set configuration values through environment variables that DuckDB respects.

So, I don't want to do this:

```
SET custom_extension_repository=getenv("DUCKDB_CUSTOM_EXTENSION_REPOSITORY");
```

I want to do this:

```
export DUCKDB_CUSTOM_EXTENSION_REPOSITORY='example'
duckdb
# Now custom_extension_repository is set to 'example'
```

This is already possible for the configuration options of the `httpfs` extension. But it would be helpful to make this possible for all configuration options in DuckDB.

**Reply by Mytherin** [reply]
The DuckDB CLI reads the `~/.duckdbrc` file on start-up and executes all commands there, so you could use the `getenv` function to accomplish exactly that in the latest master build (note that this feature is not shipped yet in v0.9.2), e.g.:

```bash
$ > cat ~/.duckdbrc
SET custom_extension_repository=getenv('DUCKDB_CUSTOM_EXTENSION_REPOSITORY');
$ > export DUCKDB_CUSTOM_EXTENSION_REPOSITORY=my_repository
$ > build/debug/duckdb                                     
-- Loading resources from /Users/myth/.duckdbrc
v0.9.3-dev927 65916b8ea6
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D SELECT * FROM duckdb_settings() WHERE name='custom_extension_repository';
┌─────────────────────────────┬───────────────┬─────────────────────────────────────────────────────────────────┬────────────┐
│            name             │     value     │                           description                           │ input_type │
│           varchar           │    varchar    │                             varchar                             │  varchar   │
├─────────────────────────────┼───────────────┼─────────────────────────────────────────────────────────────────┼────────────┤
│ custom_extension_repository │ my_repository │ Overrides the custom endpoint for remote extension installation │ VARCHAR    │
└─────────────────────────────┴───────────────┴─────────────────────────────────────────────────────────────────┴────────────┘

```

**Comment by 0xbe7a** [comment]
I think having a environment variable respected by libduckdb would be really useful. We are trying to package duckdb-extensions into conda-forge. Ideally users should be able to install extensions by installing a conda-package e.g. duckdb-s3. Installing it would result in the s3 extension being installed `$CONDA_PREFIX/duckdb`. Ideally duckdb would respect a environment variable `DUCKDB_EXTENSIONS_DIR` or similar that could be configured by conda to point to the actual folder containing all installed extensions. This way they would be manageable through the package manager and pre-installable which would make deploying them in a docker image or managed environments much easier.

**Reply by herebebeasties** [reply]
This would be great. Note that DuckDB already has `SET extension_directory = '/path/to/your/extension/directory';` so this should be `DUCKDB_EXTENSION_DIRECTORY` to match that.

**Comment by nicornk** [comment]
My +1 to this feature request. I would like to set the temp_directory to the ephemeral instance storage of the ec2 instances centrally. That way the user would not have to know about these details and could just enjoy a fast duckdb instance.

**Comment by ThomasG77** [comment]
I'm mostly happy with the `getenv`approach but I would prefer to have an optional fallback option e.g 

`SET custom_extension_repository=getenv('DUCKDB_CUSTOM_EXTENSION_REPOSITORY');` would become `SET custom_extension_repository=getenv('DUCKDB_CUSTOM_EXTENSION_REPOSITORY', 'mydefaultvalueinmyscript');`

PS: biased because of Python approach where it's possible to do the following

```python
import os

# The current getenv approach we got
custom_extension_repository = os.environ.get("DUCKDB_CUSTOM_EXTENSION_REPOSITORY")
# The getenv approach I want except applied in a DuckDb context
custom_extension_repository = os.environ.get("DUCKDB_CUSTOM_EXTENSION_REPOSITORY", 'another_var')
print(custom_extension_repository)
```

**Reply by Tishj** [reply]
That makes sense to me, you could make a PR for that

**Comment by JimCallahanOrlando** [comment]
Also please add my +1 to this feature request.
I would like to use 
-- Set DuckDB search path to OS working directory $PWD (named for pwd command in Linux/Unix)
-- This enables relative paths such as '\data\flights.csv\'
SET file_search_path = getenv('PWD');

**Reply by JimCallahanOrlando** [reply]
Unfortunately, 
SELECT getenv('PWD') ;
did not seem to work for me in Windows 10,
even though this is based on the getenv('HOME') example in the documentation (which also  did not work in Windows 10).
SELECT getenv('HOME') ;
https://duckdb.org/docs/stable/sql/functions/utility.html

But, in Windows PowerShell (PS)
echo $HOME
echo $PWD
Did work.

**Reply by Mytherin** [reply]
Are you using the CLI? `getenv` is only available in the CLI.

**Reply by JimCallahanOrlando** [reply]
Yes, DuckDB (D prompt) command line interface run from Windows 10 Power Shell.

PS D:\test_project> echo $HOME
C:\Users\JimCa
PS D:\test_project> echo $PWD

Path
----
D:\test_project


PS D:\test_project> duckdb testdb
DuckDB v1.4.0 (Andium) b8a06e4a22
Enter ".help" for usage hints.
D SELECT getenv('HOME');
┌────────────────┐
│ getenv('HOME') │
│    varchar     │
├────────────────┤
│                │
└────────────────┘
D SELECT getenv('PWD');
┌───────────────┐
│ getenv('PWD') │
│    varchar    │
├───────────────┤
│               │
└───────────────┘
D
